cpp/src/randomforest/randomforest_impl.cuh

```diff
@@ -374,49 +441,93 @@ const DecisionTree::DecisionTreeRegressor<T>* rfRegressor<T>::get_trees_ptr()
-      this->rf_params.tree_params.n_bins,
-      this->rf_params.tree_params.split_algo);
+  MLCommon::device_buffer<char>* rows_temp_storage[n_streams];
+  size_t temp_storage_bytes[n_streams];
+  for (int i = 0; i < n_streams; i++) {
+    rows_temp_storage[i] = nullptr;
+    temp_storage_bytes[i] = 0;
+    if (this->rf_params.tree_params.split_algo != SPLIT_ALGO::GLOBAL_QUANTILE) {
+      CUDA_CHECK(cub::DeviceRadixSort::SortKeys(
+        rows_temp_storage[i], temp_storage_bytes[i], selected_rows[i]->data(),
+        sorted_selected_rows[i]->data(), n_sampled_rows, 0,
+        8 * sizeof(unsigned int), stream));
+      // Allocate temporary storage
+      rows_temp_storage[i] = new MLCommon::device_buffer<char>(
+        handle.getDeviceAllocator(), stream, temp_storage_bytes[i]);
+    }
+  }
 
+  std::shared_ptr<TemporaryMemory<T, T>> tempmem[n_streams];
+  for (int i = 0; i < n_streams; i++) {
+    tempmem[i] = std::make_shared<TemporaryMemory<T, T>>(
+      local_handle[i].getImpl(), n_rows, n_cols, 1, 1,
+      this->rf_params.tree_params.n_bins,
+      this->rf_params.tree_params.split_algo,
+      this->rf_params.tree_params.max_depth);
+  }
+  //Preprocess once only per forest
   if ((this->rf_params.tree_params.split_algo == SPLIT_ALGO::GLOBAL_QUANTILE) &&
       !(this->rf_params.tree_params.quantile_per_tree)) {
     preprocess_quantile(input, nullptr, n_rows, n_cols, n_rows,
-                        this->rf_params.tree_params.n_bins, tempmem);
+                        this->rf_params.tree_params.n_bins, tempmem[0]);
+    for (int i = 1; i < n_streams; i++) {
+      CUDA_CHECK(cudaMemcpyAsync(
+        tempmem[i]->d_quantile->data(), tempmem[0]->d_quantile->data(),
+        this->rf_params.tree_params.n_bins * n_cols * sizeof(T),
+        cudaMemcpyDeviceToDevice, tempmem[i]->stream));
+      memcpy((void*)(tempmem[i]->h_quantile->data()),
+             (void*)(tempmem[0]->h_quantile->data()),
+             this->rf_params.tree_params.n_bins * n_cols * sizeof(T));
+    }
   }
+
+#pragma omp parallel for num_threads(n_streams)
   for (int i = 0; i < this->rf_params.n_trees; i++) {
-    this->prepare_fit_per_tree(handle, i, n_rows, n_sampled_rows,
-                               selected_rows.data(),
-                               sorted_selected_rows.data(),
-                               rows_temp_storage->data(), temp_storage_bytes);
+    int stream_id = omp_get_thread_num();
+    unsigned int* rowids;
+    unsigned int* selected_ptr = nullptr;
+    char* temp_storage_ptr = nullptr;
+    if (this->rf_params.tree_params.split_algo != SPLIT_ALGO::GLOBAL_QUANTILE) {
+      rowids = sorted_selected_rows[stream_id]->data();
+      selected_ptr = sorted_selected_rows[stream_id]->data();
+      temp_storage_ptr = rows_temp_storage[stream_id]->data();
+    } else {
+      rowids = selected_rows[stream_id]->data();
+    }
+    this->prepare_fit_per_tree(local_handle[stream_id].getImpl(), i, n_rows,
```

Comment:
Same as above- cumlHandle is not thread-safe so we shouldn't be accessing it from multiple threads. cudaStreams and the deviceAllocator should be thread safe.

Reply:
Done thanks